filters/auth/jwt_validation.go

```diff
@@ -0,0 +1,257 @@
+package auth
+
+import (
+	"crypto/rsa"
+	"encoding/base64"
+	"fmt"
+	"math/big"
+	"strings"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/opentracing/opentracing-go"
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/filters"
+)
+
+const (
+	JwtValidationName = "jwtValidation"
+)
+
+type (
+	jwtValidationSpec struct {
+		options TokenintrospectionOptions
+	}
+
+	jwtValidationFilter struct {
+		authClient      *authClient
+		claims          []string
+		upstreamHeaders map[string]string
+	}
+)
+
+var rsakeys map[string]*rsa.PublicKey
+
+func NewJwtValidationWithOptions(o TokenintrospectionOptions) filters.Spec {
+	return &jwtValidationSpec{
+		options: o,
+	}
+}
+
+func NewJwtValidation(timeout time.Duration) filters.Spec {
+	return NewJwtValidationWithOptions(TokenintrospectionOptions{
+		Timeout: timeout,
+		Tracer:  opentracing.NoopTracer{},
+	})
+}
+
+func JwtValidationWithOptions(
+	create func(time.Duration) filters.Spec,
+	o TokenintrospectionOptions,
+) filters.Spec {
+	s := create(o.Timeout)
+	ts, ok := s.(*tokenIntrospectionSpec)
+	if !ok {
+		return s
+	}
+
+	ts.options = o
+	return ts
+}
+
+func (s *jwtValidationSpec) Name() string {
+	return JwtValidationName
+}
+
+func (s *jwtValidationSpec) CreateFilter(args []interface{}) (filters.Filter, error) {
+	sargs, err := getStrings(args)
+	if err != nil {
+		return nil, err
+	}
+	if len(sargs) < 2 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	issuerURL := sargs[0]
+
+	cfg, err := getOpenIDConfig(issuerURL)
+	if err != nil {
+		return nil, err
+	}
+
+	var ac *authClient
+	var ok bool
+	if ac, ok = issuerAuthClient[issuerURL]; !ok {
+		ac, err = newAuthClient(cfg.JwksURI, tokenInfoSpanName, s.options.Timeout, s.options.MaxIdleConns, s.options.Tracer)
```

Comment:
The first argument `cfg.JwksURI` is wrong. JWKs URI is the one you can pull public keys from.
I would suggest to use a goroutine to refresh every 30m the public keys from the JWKs URI and below in `Request()` you just validate the token with the publickeys stored in the filter. You don't need the roundtrip to the JWKs uri in the hot path.

Reply:
This is clear in general. Will try to implement the change.

Reply:
Resolved this by providing following functionality: on filter creation I am getting oidc config with JWKs URI and scheduling the call to load keys from the uri every 30 min.